libraries/persistent_store/fuzz/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2019-2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO(ia0): Remove when used.
+#![allow(dead_code)]
+
+/// Bit-level entropy source based on a byte slice shared reference.
+struct Entropy<'a> {
+    /// The byte slice shared reference providing the entropy.
+    data: &'a [u8],
+
+    /// The bit position in the byte slice of the next entropy bit.
+    bit: usize,
+}
+
+impl Entropy<'_> {
+    /// Creates a bit-level entropy given a byte slice.
+    fn new(data: &[u8]) -> Entropy {
+        let bit = 0;
+        Entropy { data, bit }
+    }
+
+    /// Consumes the remaining entropy.
+    fn consume_all(&mut self) {
+        self.bit = 8 * self.data.len();
+    }
+
+    /// Returns whether there is entropy remaining.
+    fn is_empty(&self) -> bool {
+        assert!(self.bit <= 8 * self.data.len());
+        self.bit == 8 * self.data.len()
+    }
+
+    /// Reads a bit.
+    fn read_bit(&mut self) -> bool {
+        if self.is_empty() {
```

Comment:
The `assert` in `is_empty` is about an internal invariant. It is always possible to read a bit even when there is no more entropy. This is used by fuzzing because it works by cycles and only always finishes a cycle (loop iteration) before checking whether there is still entropy to start a new cycle (loop condition).

Reply:
If running out of entropy is expected, why not move the pointer back to the beginning, instead of using constant zeros? (Best case would probably be a pseudo random pass over the array, and resetting later.) Maybe I don't understand the point of this whole class, since I haven't seen the application yet (then it's a documentation issue most likely).
Are you just seeding something with this entropy? Why is using a built-in random generator not good enough?

Reply:
Yes that could be an option, but then a bit would be used more than once. I clarified in the top-level documentation that this doesn't happen.

Built-in random generators I've found work only at the byte level and waste a lot of information for the kind of operations I do. I still use a pseudo-random generator internally when I need byte-level input and don't want to waste entropy.

Reply:
One bit is moved more than once in any case, right? In your implementation, it's the constant zero bit.

I feel like I'm not getting the essence. What is information waste? Do you want to store the minimal amount of entropy necessary to reproduce during fuzzing? And a random seed wouldn't be an option for that?

Reply:
Taken offline, I was in fact lacking the basic understanding what this class is about :)

Resolution: The fuzzer operates on the byte slice, and the store reads its information as an encoding of operations. Please add this or a similar high-level overview to the class documentation.

Reply:
Done.